### PR TITLE
e2e: start 100 pods in the density test

### PR DESCRIPTION
**Pull request by yujuhong**

### CODE REVIEW

**Comment by yujuhong** [comment]
/cc @dchen1107 

This should pass without #21438, but #21438 will make cleanup time shorter (and better for the next test).

**Comment by gmarek** [comment]
cc @wojtek-t

File: test/e2e/density.go [hunk 1]
```diff
@@ -181,15 +187,13 @@ var _ = Describe("Density", func() {
...
-		// We might want to enable those tests in the future.
 		{podsPerNode: 50, runLatencyTest: false, interval: 10 * time.Second},
-		{podsPerNode: 100, runLatencyTest: false, interval: 1 * time.Second},
+		{podsPerNode: 100, runLatencyTest: true, interval: 10 * time.Second},
 	}
 
 	for _, testArg := range densityTests {
 		name := fmt.Sprintf("should allow starting %d pods per node", testArg.podsPerNode)
-		if testArg.podsPerNode == 30 {
+		if testArg.podsPerNode == 100 {
```
Comment (wojtek-t) [inline test/e2e/density.go]: I'm almost convinved this will not work, because IIUC we have a hard limit on MaxPods set to 100, and there are also some system pods, and additionally, we are starting some pods as part of "latency" phase. So I think this should be something less than 100 (like 95?)

File: test/e2e/density.go [hunk 2]
```diff
@@ -210,6 +214,7 @@ var _ = Describe("Density", func() {
 				PodStatusFile:        fileHndl,
 				Replicas:             totalPods,
 				MaxContainerFailures: &MaxContainerFailures,
+				Timeout:              time.Duration(math.Ceil(float64(itArg.podsPerNode)/minPodStartupThroughput)) * time.Minute,
```
Comment (wojtek-t) [inline test/e2e/density.go]: Please don't change it - it's not connected only to kubelet throughput, but also it's related to scheduler and rest of the system. And we definitely not have 15 pods/s throughput on scheduler for large clusters right now.

File: test/e2e/density.go [hunk 3]
```diff
@@ -181,15 +187,13 @@ var _ = Describe("Density", func() {
 		// TODO: Expose runLatencyTest as ginkgo flag.
 		{podsPerNode: 3, runLatencyTest: false, interval: 10 * time.Second},
 		{podsPerNode: 30, runLatencyTest: true, interval: 10 * time.Second},
-		// More than 30 pods per node is outside our v1.0 goals.
-		// We might want to enable those tests in the future.
 		{podsPerNode: 50, runLatencyTest: false, interval: 10 * time.Second},
-		{podsPerNode: 100, runLatencyTest: false, interval: 1 * time.Second},
+		{podsPerNode: 100, runLatencyTest: true, interval: 10 * time.Second},
```
Comment (wojtek-t) [inline test/e2e/density.go]: I'm almost convinved this will not work, because IIUC we have a hard limit on MaxPods set to 100, and there are also some system pods, and additionally, we are starting some pods as part of "latency" phase. So I think this should be something less than 100 (like 95?)

File: test/e2e/density.go (hunk 1 above)
Comment (wojtek-t) [inline test/e2e/density.go]: I'm wondering what's the best way to do it. Note that we are running "Feature:Performance" suite also on larger clusters (e.g. 1000-node kubemark), where we don't want to run 100 pods/node density.

So personally, I would suggest setting some other tag for 100 pods/node, e.g.:
[Feature:PerformanceWithNode]   <- please suggest better name for it
and change our jenkins script to run tests with both tags in our 100-node scalability cluster.

@gmarek - thoughts?

**Comment by wojtek-t** [comment]
@yujuhong - did you run this test on large cluster? scalability suite is blocking merge-bot so I would like to avoid breaking it :)

**Comment by wojtek-t** [comment]
Or what is probably even better, we might want to set up a separate project for testing this configuration.

File: test/e2e/density.go (hunk 3 above)
Comment (gmarek) [inline test/e2e/density.go]: Hard limit it 110 IIRC.

File: test/e2e/density.go (hunk 1 above)
Comment (gmarek) [inline test/e2e/density.go]: Yup - [Feature:HighDensityPerformance] tag is the way to go IMO.

We also should scheduler regular 300 Node/kubemark runs to make sure we're meeting our SLO in maximal setup.

Comment (wojtek-t) [inline test/e2e/density.go]: [Feature:HighDensityPerformance] - I like this.

**Comment by gmarek** [comment]
```
Feb 11 16:01:30.809: INFO: WARNING Top latency metric: {Resource:pods Verb:LIST Latency:{Perc50:47.341ms Perc90:77.505ms Perc99:1.893977s}}
Feb 11 16:01:30.810: INFO: Top latency metric: {Resource:nodes Verb:LIST Latency:{Perc50:102.932ms Perc90:186.152ms Perc99:301.703ms}}
Feb 11 16:01:30.810: INFO: Top latency metric: {Resource:replicationcontrollers Verb:PUT Latency:{Perc50:2.218ms Perc90:3.898ms Perc99:34.601ms}}
Feb 11 16:01:30.810: INFO: Top latency metric: {Resource:nodes Verb:PUT Latency:{Perc50:3.407ms Perc90:5.424ms Perc99:26.829ms}}
Feb 11 16:01:30.810: INFO: Top latency metric: {Resource:pods Verb:POST Latency:{Perc50:1.849ms Perc90:3.676ms Perc99:20.713ms}}
```

We believe that Pod List is a red herring though, but we should make sure it is.

File: test/e2e/density.go (hunk 3 above)
Comment (yujuhong) [inline test/e2e/density.go]: Yes, it's 110 so it's fine. I've run this test in my e2e cluster many times.

File: test/e2e/density.go (hunk 1 above)
Comment (yujuhong) [inline test/e2e/density.go]: What does it mean when we say we support 100 pods per node and a 1000 node cluster, when we only test the performance of 30 pods per node?

Personally I am fine with changing the name, but I'd like to know what's the "expected" density on a node in a large cluster and why it's different.

File: test/e2e/density.go (hunk 2 above)
Comment (yujuhong) [inline test/e2e/density.go]: I agree node is not the only bottleneck in a larger cluster. However, the current timeout (if not set) by default is just 5 min. This may not even be enough for a small cluster on a bad day. We *need* to set a reasonable timeout anyway, and the test is not setting one based on density and/or cluster size.

**Comment by yujuhong** [comment]
> @yujuhong - did you run this test on large cluster? scalability suite is blocking merge-bot so I would like to avoid breaking it :)

Is this the best way to test it? I don't mind starting a 100 cluster and test this change, but it seems like a hassle to ask every developer to do it every time especially the cluster size grows over time :-)
@wojtek-t, @gmarek What if I add a new test that doesn't fail or is marked flaky and see if it'll stabilize?

**Comment by yujuhong** [comment]
> We believe that Pod List is a red herring though, but we should make sure it is.

Wrong issue?

File: test/e2e/density.go (hunk 2 above)
Comment (yujuhong) [inline test/e2e/density.go]: BTW, this is 15 pods per *minute*. The apiserver qps limit in kubelet is still capped at 5 per second, so this is going to be slower for kubelet to report as the number of pods scales.

File: test/e2e/density.go (hunk 1 above)
Comment (wojtek-t) [inline test/e2e/density.go]: We don't support both 100 pods/node and 1000 nodes at the same time:
https://github.com/kubernetes/kubernetes/blob/master/docs/admin/cluster-large.md#support

Comment (yujuhong) [inline test/e2e/density.go]: Thanks for the pointer! :)

Comment (wojtek-t) [inline test/e2e/density.go]: That said - I think that we need to test 100pods/node with large clusters. All I'm saying is that probably it should be a separate suite (at least initially).

Comment (yujuhong) [inline test/e2e/density.go]: IIUC, we have a 1000-node cluster/kubemark that verifies the master components can handle 30000 pods already. In that case, for v1.2, I think we just want to test
 1. A node can handle 100 pods and can still meet the latency requirements
 2. A reasonably large cluster (e.g., 100 nodes) can handle 100 pods per node

For (1), I can simply add a new test to run in a regular suite.
For (2), I think we should change the scalability suite to run 100 pods per node. Starting a new suite doesn't help much because eventually we'll have to switch to 100 pods anyway. If you're worried about the possibility of breaking the suite, we can proceed by adding a new test that doesn't fail until it's stable. 

I'll change this PR to target (1). If you still prefer adding a new suite for (2), we can do that.

**Comment by gmarek** [comment]
Actually it's a correct issue - I pasted the results from the run of Density100 test on 300 Node cluster. Sorry I didn't make it clear.

**Comment by yujuhong** [comment]
> Actually it's a correct issue - I pasted the results from the run of Density100 test on 300 Node cluster. Sorry I didn't make it clear.

I see. Thanks for the context (and for running the test)!

**Comment by yujuhong** [comment]
I have verified that the test consistently passes in a 100-node kubemark cluster. @wojtek-t @gmarek, how do you want to proceed with this PR? 
 1. Tag this test with [Feature:Performance]. This will force kubemark and scalability suite to run the new test. 
 2. Use the regular [Serial] tag, so that the test will only be running in the serial suite.
 3. Use a new [Feature:HighDensityPerformance] tag and create a new suite.

IMO, (3) doesn't add much value because we already have scalability and kubemark suites and they should be able to test this. (2) will be a temporary solution because I think eventually we want to test this with a larger cluster. WDYT?

**Comment by wojtek-t** [comment]
@yujuhong - after more thinking of it, I would like some modification of 3 which is:
- use a new [Feature:HighDensityPerformance] tag
- enable it currently only in kubemark-100
Once we are sure it works also on Jenkins [in the past we observed differences between our dev setups and Jenkins] - we can add it also to a scalability]

Does it sound ok?

**Comment by gmarek** [comment]
We'd probably go with 3) - to save time, as it probably doubles the running time of the test.

File: test/e2e/density.go (hunk 2 above)
Comment (wojtek-t) [inline test/e2e/density.go]: Having the throughput in pods/minute is very misleading (because everything else everywhere else is in pods/second).
Can you change it to 0.25 (so that we have per-second throughput)?

File: test/e2e/density.go [hunk 4]
```diff
@@ -40,8 +40,17 @@ import (
...
+const (
+	// NodeStartupThreshold is a rough estimate of the time allocated for a pod to start on a node.
+	NodeStartupThreshold = 4 * time.Second
+
+	// minPodStartupThroughput is the allowed minimum pod startup throughput
+	// per node (pods per minute). Note that latency it takes for the nodes to
+	// report the pods are running is usually the bottleneck in pod startup
+	// because when the density (pods per node) is high. We use a rather
+	// conservative number to avoid premature timeout while waiting.
+	minPodStartupThroughput = 15.0
```
Comment (wojtek-t) [inline test/e2e/density.go]: Since is not really pod startup throughput - this is per-node pod startup throughput. Can you rename it to reflect it, e.g.:
minPerNodePodStartupThroughput
or sth like that?

**Comment by yujuhong** [comment]
> We'd probably go with 3) - to save time, as it probably doubles the running time of the test.

If we go with 3, we will create a temporary suite, and eventually switch kubemark and performance suite to run this. That seems like unnecessary churn.

File: test/e2e/density.go (hunk 4 above)
Comment (yujuhong) [inline test/e2e/density.go]: Done.

File: test/e2e/density.go (hunk 2 above)
Comment (yujuhong) [inline test/e2e/density.go]: Sure. I got this from https://github.com/kubernetes/kubernetes/blob/f23034d5dac2be305b419db55dd344fbb8738f5d/test/e2e/metrics_util.go#L449

**Comment by yujuhong** [comment]
@wojtek-t @gmarek I updated the PR based on the comments. PTAL!

Would any of you be willing to create a suite for this? Thanks!

**Comment by gmarek** [comment]
We'll do this.

File: test/e2e/density.go [hunk 5]
```diff
@@ -40,8 +40,17 @@ import (
...
 )
 
-// NodeStartupThreshold is a rough estimate of the time allocated for a pod to start on a node.
-const NodeStartupThreshold = 4 * time.Second
+const (
+	// NodeStartupThreshold is a rough estimate of the time allocated for a pod to start on a node.
+	NodeStartupThreshold = 4 * time.Second
+
+	// minPodStartupThroughput is the allowed minimum pod startup throughput
+	// per node (pods per minute). Note that latency it takes for the nodes to
```
Comment (wojtek-t) [inline test/e2e/density.go]: nit: per second

File: test/e2e/density.go [hunk 6]
```diff
@@ -40,8 +40,17 @@ import (
...
+	// report the pods are running is usually the bottleneck in pod startup
+	// because when the density (pods per node) is high. We use a rather
+	// conservative number to avoid premature timeout while waiting.
+	minPerNodePodStartupThroughput = 0.25
```
Comment (gmarek) [inline test/e2e/density.go]: 0.25 per minute seems a bit low.

File: test/e2e/density.go [hunk 7]
```diff
@@ -210,6 +221,7 @@ var _ = Describe("Density", func() {
 				PodStatusFile:        fileHndl,
 				Replicas:             totalPods,
 				MaxContainerFailures: &MaxContainerFailures,
+				Timeout:              time.Duration(math.Ceil(float64(itArg.podsPerNode)/minPerNodePodStartupThroughput)) * time.Minute,
```
Comment (wojtek-t) [inline test/e2e/density.go]: This "Timeout" field is not "timeout for starting the whole RC" - it's something completely different.
So please leave this unchanged.

BTW - event if it was a real timeout, kubelet is not a bottleneck, because when you have e.g. RC with 30000 replicas. scheduling it will take much longer than that.

So please really remove this line.

**Comment by yujuhong** [comment]
Rebased and addressed all comments.

File: test/e2e/density.go [hunk 8]
```diff
@@ -99,7 +99,7 @@ func density30AddonResourceVerifier() map[string]resourceConstraint {
 	}
 	constraints["influxdb"] = resourceConstraint{
 		cpuConstraint:    2,
-		memoryConstraint: 500 * (1024 * 1024),
+		memoryConstraint: 300 * (1024 * 1024),
```
Comment (gmarek) [inline test/e2e/density.go]: Why this change?

File: test/e2e/density.go [hunk 9]
```diff
@@ -314,30 +309,8 @@ var _ = Describe("Density", func() {
...
-				nodeNames = append(nodeNames, k)
-			}
-			sort.Strings(nodeNames)
-			for _, node := range nodeNames {
-				Logf("%v: %v pause pods, system pods: %v", node, pausePodAllocation[node], systemPodAllocation[node])
-			}
-
 			if itArg.runLatencyTest {
-				By("Scheduling additional Pods to measure startup latencies")
+				Logf("Schedling additional Pods to measure startup latencies")
```
Comment (gmarek) [inline test/e2e/density.go]: Revert this change

File: test/e2e/density.go [hunk 10]
```diff
@@ -417,7 +390,7 @@ var _ = Describe("Density", func() {
 				}
 				wg.Wait()
 
-				By("Waiting for all Pods begin observed by the watch...")
+				Logf("Waiting for all Pods begin observed by the watch...")
```
Comment (gmarek) [inline test/e2e/density.go]: Revert this change.

File: test/e2e/density.go [hunk 11]
```diff
@@ -314,30 +309,8 @@ var _ = Describe("Density", func() {
 			badEvents := BadEvents(events)
 			Expect(badEvents).NotTo(BeNumerically(">", int(math.Floor(0.01*float64(totalPods)))))
 
-			// Print some data about Pod to Node allocation
```
Comment (gmarek) [inline test/e2e/density.go]: Revert this change.

File: test/e2e/density.go (hunk 8 above)
Comment (yujuhong) [inline test/e2e/density.go]: A poor rebase. Sorry. I'll rebase again.

**Comment by yujuhong** [comment]
@gmarek, I pushed the wrong local branch. Sorry about that. I've rebased and uploaded a new commit.

**Comment by wojtek-t** [comment]
LGTM - thanks!
